File: Java/src/main/java/com/gildedrose/item_helpers/ItemFactory.java
package com.gildedrose.item_helpers;

import com.gildedrose.main.Item;
import com.gildedrose.items.AgedBrieItem;
import com.gildedrose.items.BackstagePassItem;
import com.gildedrose.items.ConjuredItem;
import com.gildedrose.items.LegendaryItem;
import com.gildedrose.items.NormalItem;

import static com.gildedrose.item_helpers.ItemHandler.validate;
import static com.gildedrose.item_helpers.ItemName.getItemName;

public class ItemFactory {

    private ItemFactory() {
    }

    public static ItemType getItem(Item item) {
        validate(item);
        ItemName itemName = getItemName(item.name);
        switch (itemName) {
            case AGED_BRIE:
                return new AgedBrieItem(item);
            case LEGENDARY:
                return new LegendaryItem(item);
            case BACKSTAGE_PASS:
                return new BackstagePassItem(item);
            case CONJURED:
                return new ConjuredItem(item);
            default:
                return new NormalItem(item);
        }
    }

}

File: php/src/ItemFactory.php
<?php

declare(strict_types=1);

namespace GildedRose;

use GildedRose\Items\Interface\ItemInterface;
use GildedRose\Items\AgedBrieItem;
use GildedRose\Items\BackstagePassItem;
use GildedRose\Items\ConjuredItem;
use GildedRose\Items\NormalItem;
use GildedRose\Items\SulfurasItem;
class ItemFactory
{
    public const AGED_BRIE_ITEM = 'Aged Brie';
    public const BACKSTAGE_ITEM = 'Backstage';
    public const SULFURAS_ITEM = 'Sulfuras';
    public const CONJURED_ITEM = 'Conjured';

    /**
     * @param Item $item
     *
     * @return ItemInterface
     */
    public static function createItem(Item $item): ItemInterface
    {
        return match(true) {
            strpos($item->name, self::AGED_BRIE_ITEM) !== false => new AgedBrieItem(),
            strpos($item->name, self::BACKSTAGE_ITEM) !== false => new BackstagePassItem(),
            strpos($item->name, self::SULFURAS_ITEM) !== false => new SulfurasItem(),
            strpos($item->name, self::CONJURED_ITEM) !== false => new ConjuredItem(),
            default => new NormalItem()
        };
    }

}
